### DISCUSSION TITLE
Database Implementation

### DISCUSSION THREAD

**Discussion by Ar2691** [uprm-inso4101-2024-2025-s1/semester-project-regiupr#4] [General]
Since we might have to wait for the department's response to integrate our system and to avoid future complications/delays 
with it I suggest we feed our system data from a database we create.

**Comment by carlospepin23** [comment]
> Since we might have to wait for the department's response to integrate our system and to avoid future complications/delays with it I suggest we feed our system data from a database we create.

To contribute to the idea of creating a database and mitigate delays, Nelson and I have in mind that it can be done this way. Please feel free to comment.

Objective:

Delay Mitigation: Reduce integration issues with the department's system and avoid development disruptions.
Standalone Data Source: Establish a standalone database that allows development to move forward and test without relying on the department system.

Description: As mentioned by AR2691, since we could face delays with the integration of the university's department system, it would be good to create a database with “dummy” information. This will allow us to continue with the development without interruptions. Also, in this way, we will be able to validate the integration and functionality of the system before completing the integration with the department's system.

SQLite recommendation: We researched and found that SQLite is integrated with Python (https://docs.python.org/3/library/sqlite3.html). Its use will make it easy to configure and manage the database without the need for additional downloads or configuration.

Schema Proposal:

Students Table: To store information about students (ID, name, email, etc.).
Courses Table: To store information about courses (name, course code, etc.).

Tasks that may arise:

-Schema Design: Create a detailed schema for the necessary tables and relations.
-Database Creation: Implement the schema in SQLite and load initial data if needed.
-Database Integration: Adjust the system to connect to and work with the SQLite database.
-Testing: Verify the performance of the database and its integration with the system, ensuring that it meets the requirements.

**Reply by nelson-escalona** [reply]
As I discussed with Carlos, this is a great start for the database implementation with creating the database scheme being one of the first tasks both of us work on.

**Comment by Yamil-Serrano** [comment]
In the students table, do you think it would be a good idea to add a list that stores all the courses that the student has already taken? It would be something like (ID, name, email, [COURSE_101, COURSE_102, COURSE_103... etc].) This way we can have a record of the courses that the student has taken and which ones they need to take according to their academic curriculum.

**Reply by nicolle-feliciano** [reply]
I think this would be a good idea since one of the things we want to fix with this app is the hassle of figuring out which courses students have unlocked. With this list, the app could show students which courses they can take based on what they've already taken.

**Comment by cristianramos9** [comment]
So SQLite3 is the database that will be used for the project. Is this correct?

**Reply by nelson-escalona** [reply]
Yes. For now, SQLite is the choice we're going for. We have a plan to implement the database in AWS for Milestone 2, just so the database is centralized. We may be changing the database to MySQL for the AWS implementation, but that's just a maybe. 

Once the DB design gets approved, the DB team will prepare instructions for developers as to how to download and install the database to their respective testing/development environment, so that every developer can run their own tests.
